### DISCUSSION TITLE
Phase 1: Schema Transformation

### DISCUSSION THREAD

**Discussion by matt-psaltis** [ChilliCream/hotchocolate#4959] [Feature Request]
# SchemaCoordinates

* Memory pooling and parent inheritance/hierarchical opportunities here to reduce object duplication across sub-nodes (fields, directives, arguments etc)
  * Referencing the original node object minimises string duplication
  * Having a parent SchemaCoordinate uses references to construct the human readable components as necessary - Provides opportunities to ignore the human readable elements unless required. Reduces memory consumption of duplicate strings at every level.
* Q: When will schema coordinates be generated in the pipeline?

# Transformation process: Inputs

An implementation of the IServiceDefinition interface will be provided to the Transformation process from the Apply Extensions phase.

```csharp
public interface IServiceReference
{
  string Name { get; }
  Uri BaseAddress { get; }
  IEnumerable<IServiceFeature> Features { get; }
}

public interface IServiceFeature
{
  string Name { get; }
  bool Enabled { get; }
  IDictionary<string, object> Settings { get; }
}
```

```csharp
public interface IServiceDefinition
{
  IServiceReference ServiceReference { get; }
  IEnumerable<DocumentNode> Documents { get; }
}
```

# Transformation process: Contracts

```csharp
public interface ISchemaTransformer
{
  ValueTask<ITransformationResult> Transform(IServiceDefinition serviceDefinition, SchemaTransformationOptions options);
}

public class SchemaTransformationOptions {
  bool AllowExtensions { get; } = true;
}
```

```csharp
public interface ITransformationResult
{
  ISchemaDocument SubGraph { get; }
  ??? Errors { get; }
}
```

**Comment by tsinghammer** [comment]
Do we want to include extensions in this discussion, too? Or create a new one?

**Reply by matt-psaltis** [reply]
Happy to include all the Phase 1 stuff in here personally. Helps to clarify where the boundaries are between the two components we added the other day to Miro.

**Comment by tsinghammer** [comment]
Regarding type extension: Consent was to do make this the very first step in the pipeline?
Do we need any annotations coming out of this step, like doe we need to track fields that came from the extended type?

**Reply by tsinghammer** [reply]
Btw. is this all already going into main as the code is not yet used?
